ratis-server/src/main/java/org/apache/ratis/server/leader/LogAppenderBase.java

```diff
@@ -158,6 +158,11 @@ private TermIndex getPrevious(long nextIndex) {
       }
     }
 
+    final long raftLogLastPurgedIndex = getRaftLog().getStartIndex() - 1;
+    if (raftLogLastPurgedIndex == previousIndex) {
+      return TermIndex.valueOf(server.getInfo().getCurrentTerm(), raftLogLastPurgedIndex);
```

Comment:
We should check term-begin-index.

Reply:
If so, shall we loose the limit the check when ```containsTermIndex```, only checks the index for the snapshot case?

Reply:
@Xushaohong , you are right that the term may be skipped so that returning `currentTerm - 1 ` is incorrect.  Good catch!  How about the following?
```java
    if (getRaftLog().getStartIndex() == nextIndex) {
      if (nextIndex > leaderState.getTermBeginIndex()) {
        return TermIndex.valueOf(server.getInfo().getCurrentTerm(), previousIndex);
      }
    }
```

Reply:
This should be fine for the case that the snapshot is installed in the current term, but we still need to solve the case ```nextIndex == beginIndex``` which is the root cause of the problem. 
The leader doesn't know the term when the follower took its snapshot since the leader's snapshot info has been overwritten and the current PB(InstallSnapshotReplyProto) has no such snapshot term clues.

We can do either of the following:
1. add the field term in PB
2. loosen the check for consistency when **containsTermIndex**,   only checks the index for the snapshot case

Reply:
> ...  the case nextIndex == beginIndex which is the root cause of the problem.

Why this is the root cause?  Note that `beginIndex` has nothing to do with log purging.

Reply:
> > ...  the case nextIndex == beginIndex which is the root cause of the problem.
> 
> Why this is the root cause? 

Sorry for the confusing words,  this is the case for what we met.  You can take a look at the scenario above.

> Note that beginIndex has nothing to do with log purging

yep, the beginIndex is just for distinguishing whether the **previous** entry happened in the current term or before.

Reply:
> ...  this is the case for what we met. ...

Are you sure?  The description did not mention that a new term had begun.  What was the beginIndex in that case?

Reply:
I checked the log again. 

Leader A's beginIndex is  **term 17 for 3407894** , raft log begin index is **log_3402406-3407893**

Follower B's snapshotIndex is (t:9, i:3402405).

So, this is the case nextIndex < leaderState.getTermBeginIndex. We are both wrong lol.

Reply:
Thanks for checking.  If it was before the term begin index, it cannot be easily fixed.

What was Leader's latest snapshot index?

Reply:
For the leader:
**when the term is 9, the close triggered a takeSnapshot (t:9, i:3407893)** this should be what you want
when the term is 17, the close triggered a  takeSnapshot (t:17, i:3413643)

Reply:
Then, did the leader send the latest snapshot (t:17, i:3413643)?

Reply:
> Then, did the leader send the latest snapshot (t:17, i:3413643)?

NO.
When leader A restarted, it remained to be the leader and tried to send its latest entries, but B does not have them and triggered Inconsistency and thus A sent the `notifyInstallSnapshot`. Since A's raft log.startIndex had not changed, A will still try to append the log starting from 3402406,  which caused the loop.

Reply:
When A sent `notifyInstallSnapshot`, did B install a snapshot by itself?

> Since A's raft log.startIndex had not changed, A will still try to append the log starting from 3402406, which caused the loop.

This looks correct -- the Leader A was sending heartbeats.  It is supposed to work like that --  A will keep sending heartbeats until B has installed a snapshot.

Reply:
> > Since A's raft log.startIndex had not changed, A will still try to append the log starting from 3402406, which caused the loop.
> 
> This looks correct -- the Leader A was sending heartbeats. It is supposed to work like that -- A will keep sending heartbeats until B has installed a snapshot.

Just to clarify the scenario:
A's notifyInstallSnapshot tells B to install at its first available Term index (t:9, i:3402406)
B returns ALREADY_INSTALLED, current snapshot index: 3402405
Then A sends entries starting from 3402406, but A cannot find the **previous entry** 3402405 from the raft log or from the latest SM's snapshot.  

The reason is A's snapshot info is not the one that causes the purging raft log to 3402405.
 According to the log, A's snapshotIndex should be (t:9, i:3407893), which does not equal 3402405.
No matter how many times A restarts, as long as A's raft log's start index is not updated, the problem will remain. 
So the follower is not working. I am afraid by default we do not use **purgeUptoSnapshotIndex**, the purgeIndex will not be updated.

Reply:
[#89 ](https://github.com/hanishakoneru/ratis/blob/d75cab2b5c44b6b321b98b812df8919be169fea5/ratis-server/src/main/java/org/apache/ratis/server/impl/LogAppender.java#L251-L252)

I found a historical issue here, but I am not sure why only ignores this case in the appender rather than also in the receiver.
If we inherit the logic here, shall we soften the limit in the follower's check?

Reply:
It sounds good.  Let's try it.